cpp/include/raft/detail/layout_padded_general.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cudart_utils.h>
+#include <raft/thirdparty/mdspan/include/experimental/mdspan>
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+
+namespace std {
+namespace experimental {
+
+enum class StorageOrderType { column_major_t, row_major_t };
+
+template <class Extents>
+MDSPAN_INLINE_FUNCTION constexpr auto padded_row_major_strides(size_t ByteAlignment,
+                                                               size_t valueTypeSize,
+                                                               Extents const& __exts)
+  -> std::array<size_t, Extents::rank()>
+{
+  auto alignment                              = std::max<size_t>(ByteAlignment / valueTypeSize, 1);
+  std::array<size_t, Extents::rank()> strides = std::array<size_t, Extents::rank()>{};
+  size_t stride                               = 1;
+  for (size_t r = Extents::rank() - 1; r > 0; r--) {
+    strides[r] = stride;
+    if (stride == 1) {
+      stride *= std::max<size_t>(alignment, raft::alignTo(__exts.extent(r), alignment));
+    } else {
+      stride *= __exts.extent(r);
+    }
+  }
+  strides[0] = stride;
+  return strides;
+}
+
+template <class Extents>
+MDSPAN_INLINE_FUNCTION constexpr auto padded_col_major_strides(size_t ByteAlignment,
+                                                               size_t valueTypeSize,
+                                                               Extents const& __exts)
+  -> std::array<size_t, Extents::rank()>
+{
+  auto alignment                              = std::max<size_t>(ByteAlignment / valueTypeSize, 1);
+  std::array<size_t, Extents::rank()> strides = std::array<size_t, Extents::rank()>{};
+  size_t stride                               = 1;
+  for (size_t r = 0; r + 1 < Extents::rank(); r++) {
+    strides[r] = stride;
+    if (stride == 1) {
+      stride *= std::max<size_t>(alignment, raft::alignTo(__exts.extent(r), alignment));
+    } else {
+      stride *= __exts.extent(r);
+    }
+  }
+  strides[__exts.rank() - 1] = stride;
+  return strides;
+}
+
+// similar to layout_strided, but contiguous with padding in second smallest stride dimension
+template <typename ValueType,
+          StorageOrderType StorageOrder = StorageOrderType::row_major_t,
+          size_t ByteAlignment          = 128>
+struct layout_padded_general {
```

Comment:
@mhoemmen, sorry to call you out again, could you have a quick look at our attempt to define the padded layout - whether it is properly in the spirit of the mdspan spec?

Reply:
@achirkin I'm happy to help out! : - )

My main issue with this design is that the layout mapping cannot possibly be responsible for alignment.  The layout can do padding to a particular size, but it can't control whether the pointer is aligned (so that you can do aligned SIMD loads, etc.).  This is because the layout never sees the pointer; it's the accessor that sees the pointer.

Would you instead consider the following design?

* The layout mapping 
    * pads each row (if row-major, or column, if column-major) to a given number of _elements_ (not a number of bytes); and
    * encodes the padding as a `size_type` or `integral_constant<size_type, Value>` template parameter.
* The accessor
    * encodes the alignment as a `size_t` or `integral_constant<size_t, Value>`;
    * defines the `pointer` type with the appropriate attributes to tell the compiler that it is aligned; and
    * has a different `offset_policy` (e.g., `default_accessor`) to express that offset access may not be aligned.

A type alias and possibly also a helper create-an-mdspan function would then use the `value_type` to compute the padding and alignment at compile time.  For example:
```c++
// Specialize this for different value types as needed.
// The value can be size_t to simplify specialization;
// layout_padded_general can convert it to size_type.
template<class ValueType>
struct padding {
    static constexpr size_t value = 128 / sizeof(ValueType);
};

template<class ElementType, class Extents, StorageOrderType order>
using aligned_mdspan = mdspan<ElementType, Extents, layout_padded_general<order, padding_v<remove_cv_t<remove_reference_t<ElementType>>>>, aligned_accessor<ElementType, alignment_v<remove_cv_t<remove_reference_t<ElementType>>>>>

template<class ElementType, class Extents, StorageOrderType order>
aligned_mdspan<ElementType, Extents, order>
make_aligned_mdspan(ElementType* input_pointer, Extents e, StorageOrderType /* order */)
{
    using value_type = remove_cv_t<remove_reference_t<ElementType>>;
    using pointer = typename aligned_accessor<ElementType, alignment_v<value_type>>::pointer;

    // Any run-time check for alignment would happen here.
    // Taking the alignment value as integral_constant permits interesting checks,
    // like wrapping ElementType* in a proxy pointer type
    // that encodes alignment as a non-type template parameter.
    pointer aligned_pointer = 
        assert_aligned(input_pointer, integral_constant<size_t, alignment_v<value_type>>{});
    using mapping = layout_padded_general<order, padding_v<value_type>>::template mapping<Extents>;
    return {aligned_pointer, mapping{e}};
>, 
```